### PR TITLE
feat: Sparo checkout support --to/--from option for direct projects selection

### PR DESCRIPTION
**Pull request by jzhang026**
### Basic Checks

Have you run `rush change` for this change?

- [x] Yes
- [ ] No

If **No**, please run `rush change` before, this is necessary.

If adding a **new feature**, the PR's description includes:

- [x] Reason for adding this feature
Instead of creating Sparo Profiles, sometimes I just want to checkout some projects randomly based on [RushJS selector](https://rushjs.io/pages/developer/selecting_subsets/). 
- [x] How to use
`sparo checkout --to project-A project-B --from project-C project-D`
- [ ] A basic example

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

### CODE REVIEW

File: common/changes/sparo/feat-checkout-to-selector_2024-04-25-14-16.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "sparo",
+      "comment": "[sparo checkout] support cli --to/--from option for projects seletion",
```
Comment (octogonz) [inline common/changes/sparo/feat-checkout-to-selector_2024-04-25-14-16.json:5]:
```suggestion
      "comment": "[sparo checkout] support cli --to/--from option for projects selection",
```

File: build-tests/sparo-real-repo-test/etc/checkout-to.txt [hunk 2]
```diff
@@ -0,0 +1,35 @@
+Running "sparo checkout --to sparo-output-test":
```
Comment (octogonz) [inline build-tests/sparo-real-repo-test/etc/checkout-to.txt:1]: If the CLI is changing, we need to update the corresponding website docs:
https://tiktok.github.io/sparo/pages/commands/sparo_checkout/

File: build-tests/sparo-output-test/etc/checkout-help.txt [hunk 3]
```diff
@@ -21,3 +21,5 @@ Options:
                      already exists, reset it to <start-point>         [boolean]
       --profile                                            [array] [default: []]
       --add-profile                                        [array] [default: []]
+      --to                                                 [array] [default: []]
+      --from                                               [array] [default: []]
```
Comment (octogonz) [inline build-tests/sparo-output-test/etc/checkout-help.txt:25]: How do `--to` and `--from` interact with `--profile`?  Is combined with the profile's selection, to produce the union?

File: apps/sparo-lib/src/services/SparoProfileService.ts [hunk 4]
```diff
@@ -227,10 +227,14 @@ ${availableProfiles.join(',')}
    */
   public async syncProfileState({
```
Comment (EscapeB) [inline apps/sparo-lib/src/services/SparoProfileService.ts:228]: If we use sparo profile in checkout, the profile will be recorded, and after users checkout to another branch, they will use the same profile in that branch.

Should we recored from/to selector here?

Comment (jzhang026) [reply apps/sparo-lib/src/services/SparoProfileService.ts:258]: good one! let update it accordingly

File: build-tests/sparo-output-test/etc/checkout-help.txt (hunk 3 above)
Comment (jzhang026) [reply build-tests/sparo-output-test/etc/checkout-help.txt]: updated

File: build-tests/sparo-real-repo-test/etc/checkout-to.txt (hunk 2 above)
Comment (jzhang026) [reply build-tests/sparo-real-repo-test/etc/checkout-to.txt:1]: updated

**Comment by chengcyber** [comment]
I think the most breaking change from this PR is mental model, which is how we treat "sparo profile"?

Before this PR: "profile" is the only core concept in sparo commands. All commands will take care of profile, which mentioned in the discussion. For example, "sparo checkout <branch>" will sync up the folders by the recorded profiles after checking out to the target branch to make everything consistency and prohibit footguns.
At the same time, "--to" and "--from" are only supported in "sparo-ci checkout", which means the flexibility is allowed in CI.

After this PR: "profile" is changed. It's now more like a convenience way to "--to" and "--from". It unveils powerful "--to" and "--from" and more free to let users decide how to use it. As you can tell, the power is a double-edge sword. It can bring easy-to-use story and it can also make it pretty error-prone.
The following design questions would be:
1. Do we need to record "--to" and "--from" to make project folders correct consistency after running "sparo checkout", "sparo pull" ... If so, it's pretty hard to maintain comparing with "profile"... 
2. If we are recording "--to" and "--from", we need a way to clear them or reset them. Just like the idea of "--profile" resets the current profiles and "--add-profile" appends more profiles.
3. We might also need the unimplemented command "sparo inspect" to help people understand the current state. "sparo inspect" can display both the recorded profiles and "--to" and "--from" parameters.


My proposal:
I do think some of our users could be an "advance" user which requests more freedom way to use "sparo" and they can take responsibility to make it work well. So, A much more balance implementation could be introducing a feature toggle to control the behavior of "sparo checkout". To be more specific, after running `sparo config sparo.checkout advance`, "sparo" now can allow the "--to" and "--from" parameter used in "sparo checkout". And we treat those "--to" and "--from" as a temporary state and doesn't make them into consideration when running "sparo checkout <new_branch>".

**Review by chengcyber — APPROVED** [review]

**Pull request merged**